pinot-core/src/main/java/com/linkedin/pinot/core/operator/docvalsets/TransformBlockValSet.java

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.operator.docvalsets;
+
+import com.clearspring.analytics.util.Preconditions;
+import com.linkedin.pinot.common.data.FieldSpec;
+import com.linkedin.pinot.core.common.BaseBlockValSet;
+import com.linkedin.pinot.core.common.BlockValSet;
+
+
+/**
+ * This class represents the BlockValSet for a TransformBlock for a specific column.
+ * It provides api's to access data specific to the column.
+ */
+public class TransformBlockValSet extends BaseBlockValSet {
+
+  protected final Object _values;
+  protected int _numDocs;
+  private final FieldSpec.DataType _dataType;
+
+  /**
+   * Constructor from raw values.
+   *
+   * @param values Block Values
+   * @param numDocs Number of docs in the block val set
+   * @param dataType Type of data
+   */
+  public TransformBlockValSet(Object values, int numDocs, FieldSpec.DataType dataType) {
+    _values = values;
+    _numDocs = numDocs;
+    _dataType = dataType;
+  }
+
+  /**
+   * Constructor from another block val set (e.g. projection block val set).
+   *
+   * @param blockValSet Another block val set from which to build this object
+   * @param numDocs Number of docs in the block val set
+   * @param dataType Type of data
+   */
+  public TransformBlockValSet(BlockValSet blockValSet, int numDocs, FieldSpec.DataType dataType) {
+    _numDocs = numDocs;
+    _dataType = dataType;
+
+    switch (dataType) {
```

Comment:
The value fetch should not happen here. You should cache the BlockValSet and fetch different types accordingly (data type and fetch type could be different).